test-harness/specs/security-scheme-validation-AC-3_failing.oas2.txt

```diff
@@ -0,0 +1,31 @@
+====test====
+Making a request (and not including any credentials) to an endpoint,
+which security depends on global security definitions,
+results in a 401 status code response.
+====spec====
+swagger: '2.0'
+paths:
+  /todos:
+    get:
+      responses:
+        200:
+          description: OK
+          schema:
+            type: number
+securityDefinitions:
+  BasicAuth:
+    type: basic
+  ApiKeyAuth:
+    type: apiKey
+    in: header
+    name: X-API-Key
+security:
+  - BasicAuth: []
+  - ApiKeyAuth: []
+====server====
+mock -p 4010
+====command====
+curl -sIXGET http://localhost:4010/todos
+====expect====
+HTTP/1.1 401 Unauthorized
+WWW-Authenticate: Basic
```

Comment:
I'm not sure about that expectation. Reading
https://tools.ietf.org/html/rfc7235#section-4.1 looks like we may want to return more than one schema since we define two ways to authenticate (?)

Reply:
@chris-miaskowski, what about:
```
WWW-Authenticate: Basic realm="*", X-API-Key realm="*"
```
cc @XVincentX, @philsturgeon

Reply:
@chris-miaskowski, that's why I ask, couldn't find anything that would tell for sure. https://tools.ietf.org/html/rfc7235 doesn't have answers. Do you know where to look?

Reply:
X-API-Key is an arbitrary header name. It could be Chunkey-Monkey or anything else. You won’t find any standards for it. 

The question is: why is real there in the first place? All of these tests should come from examples in the OpenAPI spec and/or HTTP spec but we can’t guess at how these security schemas work.

Reply:
hmm.. @philsturgeon, I'm not sure I follow. `realm` is required for `Basic` authentication scheme:
```
The Basic authentication scheme utilizes the Authentication Framework
   as follows.

   In challenges:

   o  The scheme name is "Basic".

   o  The authentication parameter 'realm' is REQUIRED ([RFC7235],
      Section 2.2).
```
that's why it's there(?). And about `X-API-Key`, from https://tools.ietf.org/html/rfc7235:
```
 WWW-Authenticate: Newauth realm="apps", type=1,
                       title="Login to \"apps\"", Basic realm="simple"
```
`Newauth` is clearly custom and it still has `realm` (although this might not be necessary) so I added `realm` for `X-API-Key` here. And why is this here? Well, I thought before implementing the security, it would be good if the tests would test the right thing (?) So I get that `realm` is not to be added for custom authentication schemes

Reply:
@lag-of-death you are right about the Basic as far I can tell. Not sure what @philsturgeon means. 

I would ignore the `realm` for anything else than `Basic`.

Reply:
I’m not sure what is causing confusion here? I asked what the source of these examples were, then Mateus responded with his examples. That’s what I asked for. 

I think the Newauth is a bit different from the sort of usage being expected in X-RANDOM-KEY and that is different again from OpenID, so as Chris said, probably ditch realm for these things. As I said this seems like some guesswork was involved, educated guesswork but still potentially not correct.

Reply:
To provide more context, there are a few potential sources you could be getting these examples from. The OpenAPI spec, the HTTP spec (which might cover things that OpenAPI cannot), or Swagger.io documentation: https://swagger.io/docs/specification/authentication/

Extra context: there are "authentication schemes" which are official specified things which go in the `Authorization` header, and those usually have a corresponding keyword which goes into the content of the `Authorization` header value. So the Basic scheme looks like `Authorization: Basic foo` and the Bearer scheme goes into `Authorization: Bearer foo`. That's all part of the Authorization header. Some good docs on that are over here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization

API keys could use the Authorization header and have an Authorization scheme, _or_ they could be totally arbitrary headers, query string parameters, or even a body param! 

A header example:

```
X-API-Key: foo
Some-Nonsense-Key: foo
Feet: foo
```

These are all valid.

```
components:
  securitySchemes:
    BasicAuth:
      type: http
      scheme: basic
    BearerAuth:
      type: http
      scheme: bearer
    NonsenseKeyAuth:
      type: apiKey
      in: header
      name: Feet
    ApiKeyAuth:
      type: apiKey
      in: header
      name: Feet
    FeetAuth:
      type: apiKey
      in: header
      name: Feet
```

It might not be a header at all:

```
http://example.org/foo?someNonsenseKey=foo 
```

This would be defined like this:

```
    SomeNonsenseAuth:
      type: apiKey
      in: query
      name: someNonsenseKey
```

There are a lot of options here. apiKey type params will not have a real. 

Understanding this stuff is a fair bit of work, I do not remember everything off the top of my head but I do have a fair bit of context and can help dig in. I am here for questions.

Reply:
`realms`s from `X-API-Key` removed. I've left `realm` here though: https://github.com/stoplightio/prism/blob/266e90c700e12529be507489682c3681c53f70b3/test-harness/specs/security-scheme-validation-AC-13_failing.only_oas3.txt#L28
what do you @philsturgeon think? should this cookie `realm` also be removed?

Reply:
That's fine, check it out: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate 

WWW-Authenticate is allowed to have a real, its in the spec there, so you're all good.